wodles/gcloud/integration.py

```diff
@@ -60,84 +62,77 @@ def get_subscription_path(self, project: str, subscription_id: str) \
-            response = self.subscriber.pull(self.subscription_path,
-                                            max_messages=max_messages,
-                                            return_immediately=True)
-        except google_exceptions.DeadlineExceeded:
-            logger.warning('Deadline exceeded when pulling messages. '
-                           'No more messages will be retrieved on this '
-                           'execution')
-            response = pubsub.types.PullResponse()
-
-        return response
+            response = self.subscriber.pull(
+                request={'subscription': self.subscription_path,
+                         'max_messages': max_messages}
+            )
+        except google.api_core.exceptions.DeadlineExceeded:
+            return 0
+
+        ack_ids = []
+        for received_message in response.received_messages:
+            formatted_message = self.format_msg(received_message.message.data)
+            logger.debug(f'Processing event: {formatted_message}')
+            ack_ids.append(received_message.ack_id)
+            self.send_message(formatted_message)
+
+        ack_ids and self.subscriber.acknowledge(
+            request={'subscription': self.subscription_path, 'ack_ids': ack_ids}
+        )
+
+        return len(response.received_messages)
 
     def process_messages(self, max_messages: int = 100) -> int:
         """Process the available messages in the subscription.
         :param max_messages: Maximum number of messages to retrieve
         :return: Number of processed messages
         """
-        processed_messages = 0
-        response = self.pull_request(max_messages)
-        while len(response.received_messages) > 0 and processed_messages < max_messages:
-            for message in response.received_messages:
-                message_data: bytes = message.message.data
-                if logger.getEffectiveLevel() == logging.DEBUG:
-                    logger.debug(f'Processing event:\n{self.format_msg(message_data.decode(errors="replace"))}')
-                self.process_message(message.ack_id, message_data)
-                processed_messages += 1  # increment processed_messages counter
-            # get more messages
-            if processed_messages < max_messages:
-                response = self.pull_request(max_messages - processed_messages)
-        return processed_messages
+        try:
+            self.wazuh_queue.connect(common.ANALYSISD)
+            with self.subscriber:
+                processed_messages = 0
+                pulled_messages = self.pull_request(max_messages)
+                while pulled_messages > 0 and processed_messages < max_messages:
+                    processed_messages += pulled_messages
+                    # get more messages
+                    if processed_messages < max_messages:
+                        pulled_messages = self.pull_request(max_messages - processed_messages)
+                return processed_messages
```

Comment:
According to [pubsub's documentation](https://cloud.google.com/pubsub/docs/reference/rpc/google.pubsub.v1#pullresponse) the response to the [pull method](https://googleapis.dev/python/pubsub/latest/subscriber/api/client.html#google.cloud.pubsub_v1.subscriber.client.Client.pull) (which is being used underneath `self.pull_request`) can be empty if there are no more messages available in the backlog.
With this design, if this method is called with let's say `max_messages=1000`, and there are only 50 messages in the backlog, the pull method would be called continuously until it gets the desired amount.

Since this script will be called with the same `max_messages` parameter after a period specified in the module's configuration, I think that this implementation and a poor configuration of the module could lead to a serious problem.
It's possible that if the number of logs produced is lesser than expected, a new instance of the program spawns before the last one ended. Then we would have two instances of the program making concurrent requests to the Google Cloud and wasting machine resources. If this process is repeated multiple times, it could potentially lead to the program crashing.

This could be solved by breaking out of the loop if the total amount of messages pulled in the last request is either zero or a close number.